### DISCUSSION TITLE
Deprecate offline session preloading

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#25301] [Archive]
The new default of Keycloak load offline sessions on demand, and not loading them at startup any more. The old code to preload them at startup should be removed.

The old default of preloading is still available and tested. This makes the code unnecessarily complex and more difficult to test. With a lot of offline token it also doesn't scale.

We also want eventually use the table structure and lookup functionality to work for all sessions, so this would be a training execerise.

See issue https://github.com/keycloak/keycloak/issues/25300 for the follow-up task.

We would like to get feedback on this. Thanks!

**Comment by sschu** [comment]
It's really hard to imagine a use case where all sessions really need to be in memory continuously and preloading them all at once is more efficient than loading them lazily. So removing this is totally fine for me.

**Comment by danielFesenmeyer** [comment]
Just one question. With the plans described above, how will an offline session be different from a normal session? Couldn't it just be represented as a normal session that never expires?

**Reply by sschu** [reply]
It is still stored persistently in contrast to normal sessions. But the direction I see this going to is to not have a differentiation because normal sessions will also be persisted in the future and then it is actually the same. I could imagine that most people will then not have offline tokens anymore but long-lasting sessions, because that is what they want in mobile apps for example. Offline tokens might be implemented in a different way than now also removing the binding of the offline token to the user session the token was originally created in.

**Reply by ahus1** [reply]
Eventually we want want to have cheap never-expiring session, still we need to figure out how a solution could look like and how we would get there. This is one step in a direction that we think would take us closer. 

Current problems we see with offline sessions: They have a last refresh timestamp, which is expensive to update. Committing something to a database costs IOPS which are expensive both on site an in the cloud. There can be congestions. There is already code which batches those updates, but some databases are not happy with that when there are concurrent updates and expiring sessions at the same time and report deadlocks. 

Current problems with all user sessions and client sessions in memory: The take a lot of memory, and when searching for something, we're iterating over all sessions, so listing sessions for a user either in the UI or for ensuring that a user is logged in only once is expensive. 

We'll take one step at a time, observe and lerne, see what works, and adjust what doesn't. So I ask for patience and feedback.

**Comment by ahus1** [comment]
Based on the feedback we received here, we're now proceeding with deprecating preloading for KC24
